src/server/game/Anticheat/AnticheatMgr.cpp

```diff
@@ -0,0 +1,453 @@
+        return true;
+    }
+
+    return false;
+}
+
+void AnticheatMgr::StartHackDetection(Player* player, MovementInfo const& movementInfo, uint16 opcode)
+{
+    if (!m_enabled)
+        return;
+
+    // Disable anticheat for Game Masters
+    if (player->CanBeGameMaster())
+        return;
+
+    AnticheatData& plData = m_playerMap[player->GetGUID()];
+    if (player->IsInFlight() || player->GetTransport() || player->GetVehicle())
+    {
+        plData.SetLastOpcode(opcode);
+        return;
+    }
+
+    for (AnticheatCheck* check : m_checks)
+        if (check->OnCheck(player, movementInfo, opcode))
+            check->HackReport(player);
+
+    plData.SetLastOpcode(opcode);
+}
+
+void AnticheatMgr::HandlePlayerLogin(Player* player, PreparedQueryResult result)
+{
+    AnticheatData& plData = m_playerMap[player->GetGUID()];
+
+    if (result)
+        plData.SetDailyReportState(true);
+}
+
+void AnticheatMgr::HandlePlayerLogout(ObjectGuid const& guid)
+{
+    // We must also delete it at logout to prevent have data of offline players in the db when we query the database (IE: The GM Command)
+    PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_DEL_AC_PLAYER_REPORTS);
+    stmt->setUInt32(0, guid.GetCounter());
+    CharacterDatabase.Execute(stmt);
+
+    // Delete not needed data from the memory.
+    m_playerMap.erase(guid);
+}
+
+void AnticheatMgr::SavePlayerData(ObjectGuid const& guid, SQLTransaction& trans)
+{
+    AnticheatData const& playerData = m_playerMap[guid];
+    PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_DEL_AC_PLAYER_REPORTS);
+    stmt->setUInt32(0, guid.GetCounter());
+    trans->Append(stmt);
+
+    uint8 index = 0;
+    stmt = CharacterDatabase.GetPreparedStatement(CHAR_INS_AC_PLAYER_REPORTS);
+    stmt->setUInt32(index  , guid.GetCounter());
+    stmt->setFloat (++index, playerData.GetAverage());
+    stmt->setUInt32(++index, playerData.GetTotalReports());
+    for (uint8 i = 0; i < MAX_REPORT_TYPES; ++i)
+        stmt->setUInt32(++index, playerData.GetTypeReports(static_cast<ReportTypes>(i)));
+
+    stmt->setUInt32(++index, playerData.GetCreationTime());
+    trans->Append(stmt);
+}
+
+bool AnticheatMgr::AnticheatGlobalCommand(ChatHandler* handler)
+{
+    SQLTransaction trans = CharacterDatabase.BeginTransaction();
+    // Mutex protected part
+    {
+        boost::shared_lock<boost::shared_mutex> lock(*HashMapHolder<Player>::GetLock());
+
+        HashMapHolder<Player>::MapType const& m = ObjectAccessor::GetPlayers();
+        for (auto const& pair : m)
+            SavePlayerData(pair.second->GetGUID(), trans);
+    }
+    CharacterDatabase.CommitTransaction(trans);
+
+    QueryResult resultDB = CharacterDatabase.Query("SELECT guid,average,total_reports FROM players_reports_status WHERE total_reports > 0 ORDER BY average ASC LIMIT 5");
```

Comment:
Yes, however it'll lead to not-updated-enough data, maybe split this command in two: anticheat save and anticheat global?